server/middleware.go

```diff
@@ -193,3 +193,72 @@ func WithCloseHandler(h ContextHandler) ContextHandler {
 		h.ServeHTTPContext(ctx, w, r)
 	})
 }
+
+// Key to use when setting the request ID to a context
+type ctxRequestIDKey int
+
+const (
+	// RequestIDHeader is the request ID key set to http headers
+	RequestIDHeader                 = "X-Request-Id"
+	// RequestIDKey is the key used to stash the request ID into the context
+	RequestIDKey    ctxRequestIDKey = 0
+)
+
+// AppIDHandler is a middleware that, if the $X-Request-ID$ header is not already
+// set on the request, generates an ID using the provided IDer then sets it on the
+// request and response. In both cases the value of the $X-Request-ID$ header is
+// set to the context
+func AppIDHandler(next http.Handler, idGen IDer) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		var err error
+		var requestID = r.Header.Get(RequestIDHeader)
+		if requestID == "" {
+			requestID, err = idGen.ID()
+			if err != nil {
+				// TODO gah, what then? I have good mind to eliminate this err altogether
```

Comment:
I think this is fine, we don't want to set the request ID to empty string, better to not set it at all.

Reply:
Agreed, but a) I shouldn't have let this TODO slip into the commit, and b) I think of this code as being important but not critical, and the idea that something that isn't critical blitzing stuff that is annoys me. In this case, this err is here solely because the UUID generator.

If you think this is fine I'll remove the TODO and leave the return.